### ISSUE TITLE
Gtk+ list box select item/child problems (Windows)

### ISSUE THREAD

**Issue by DrOli** [vmagnin/gtk-fortran#108]
We have some Gtk+ (2) code on Windows that performs a number of things, including interfacing with listboxes.  Everything works as expected (display, user selection, etc.), except when attempting to select a list item/child programmatically.  The nature of the "errors" vary depending on the exact function used (e.g. ...select_item vs. ...select_child), and also vary depending on the coding language.

For example, having made a signal connection as per:

    Call g_signal_connect( ListBox_cPtr, "select-child"//c_Null_Char, &
                            c_FunLoc(signal_SelectChild_ListBox_Smpl), &
                            c_Null_Ptr)

where "select-child" is the a "gtk native" signal (i.e. not something we created).

Then, something like:

    i = gtk_list_child_position (ListBox_cPtr, ListItem_Selected_cPtrX);

works correctly, but, something like:

    Call gtk_list_select_child (ListBox_cPtr, ListItem_Selected_cPtrX);

produces

    (ARTExGTK_D32.exe:2564): GLib-GObject-←[1;33mWARNING←[0m **: gsignal.c:3183: signal id '321' is invalid for instance '0022F054'

A similar experiment with:

    Call gtk_list_select_item( ListBox_cPtr, 1_c_Int);

causes a SIGSEGV/crash 

We know the list to be longer than 2 items, we know all the pointers are !Null, and valid.  Once, again, all other aspects of the listbox work correctly, it is only the attempt to select a list item/child programmatically that causes problems.

We believe the problem is related to Gtk's "internal issuance of a signal".  Our understanding is that s/r's like `gtk_list_select_child()` emit (in this case)  a "select-child" signal in order to simulate the user having physically selected the item/child.  Somewhere along the way, Gtk's "internal issuance" seems to loose track of the g_signal connections that exist and work with everything else.  Our understanding is that "internal signal issuance" is produced internally via one of the `Gtk/g_signal_emit()` s/r's.

Any thoughts would be appreciated.

**Comment by vmagnin** [comment]
No idea. But I read in the documentation that "gtk_list_select_child is deprecated and should not be used in newly-written code":
https://developer.gnome.org/gtk2/stable/GtkList.html#gtk-list-select-child
and more generally "GtkList has been deprecated since GTK+ 2.0 and should not be used in newly written code. Use GtkTreeView instead. "

**Comment by DrOli** [comment]
Many thanks for the prompt response.  Yes, we know it is deprecated.  Alas, we are obliged to rely GTK+ (2) due our reliance on parts of GTKExtra (which is GTK2), and GTK won't permit "mixing 2 and 3".

We were further influenced to use these constructs due to other considerations which I leave for an ASIDE I'll provide if requested, but which your are welcome to ignore..

As it happens, though our project must be in Fortran for various reasons, we do run a "mini/parallel C version" of Gtk developments, for a variety of reasons.  One of those reasons is to check if something that failed on the Fortran-side, is a "Fortran only issue" or not.

In this case, it seems to be a Fortran only issue.  When we run our C implementation, both

    gtk_list_select_child (ListBox_cPtr, ListItem_Selected_cPtrX);

    gtk_list_select_item( ListBox_cPtr, 1);

... work.

Whereas neither works on the Fortran-side, either giving the error as stated earlier for the "select_child", or with the "select_item" call, a SIGSEGV/Crash.  

Since all of the other s/r's in this family of s/r's (that we use) work, and the ones that fail seem to have in common the special trait that they actually issue an "internal select signal" to simulate the effect of a user physically clicking on the list item, via a `g_signal_emit()` call, we investigated possibilities along these lines.

.. we checked the Gtk etc source code, and indeed, these "select" s/r's are just wrappers for the `g_signal_emit()` variations.

Our gut feeling is that the internal "emit" somehow losses track of the pointers etc of the previously established `g_signal_connect's`, "id's", etc. on the Fortran-side for reasons we don't understand (at least that is one implication of the error messages, when there are some).

We tried to use explicitly `g_signal_emit()` and its variations, but we are just too inexperienced with Gtk etc, and the documentation is almost useless.  We think, that if we can figure out how/what the "signal id's", and "signal type's" etc are, as required for `g_signal_emit()` and its variations, we may be able to get the Fortran-side work, since we could then be sure it is sending the right signal id's/types/pointers, etc.

We have performed some (blind leading the blind) "experiments" both on the Fortran-side, and also writing some extra C into the source and recompiling Gtk to provide additional "Fortran interface/portals" s/r's, but no joy.  You are welcome to the details, but its a bit of a tome.

... any assistance (with anything), or explaining or using `g_signal_emit()` etc  would be much appreciated (e.g. how to write the Call to the emir s/r code correctly, compared to the known to be duff example at the start of the OP).

Of course, one alternative is to bite the bullet and build the "bicycle" that we need using the "space shuttle" ruinously expensive Gtk tools (TreeView, ListStore, etc.) ... unfortunately there is no guarantee that the Fortran-side would be any more successful with those "select/emit" constructs either.

**Comment by DrOli** [comment]
oops, I tried to edit my last post, but was not able to, so I add it here.

The `g_signal_emit` solution and the notion that the Fortran-side "signal ID's/type/Pointers" are central to this matter can be seen in the Gtk error message in the OP.  Notably, it is "unhappy" about "ID 321" trying to access "instance 0022F054".  As we don't know how to figure out the "ID's", and we know that the callback function, the listbox widget, etc etc all have pointers different from that Gtk's "internal emit" is trying to use ... which of course would explain why Gtk's "internal emit" is not able to "select" the item/child as requested.  Why, on the Fortran-side, this occurs we don't know, but perhaps some experimenting with `g_signal_emit` would help ... if we only knew how to use it :-)

**Comment by DrOli** [comment]
We have discovered something that may at least help to "light the way".  

Earlier we stated that 

    Call gtk_list_select_item( ListBox_cPtr, 1_c_Int);

caused a SIGSEGV with Gtk-Fortran, but worked on the same box with "Gtk-C" (i.e. native Gtk).

The SIGSEGV was quite nasty and produced almost no diagnostics, save for indicating that happened somewhere in `gdk_selection_property_delete()`, which is inside `libgdk-win32-2.0-0.dll`

We decided to experiment with various "builds" of Gtk+.  We found one build set that did NOT cause a SIGSEGV (this is a set that we did not build ourselves, but just dloaded the binaries).  Instead of a crash, it threw the error:

    Gtk-←[1;35mCRITICAL←[0m **: gtk_list_select_item: assertion 'GTK_IS_LIST (list)' failed

The "list" sent as ListBox_cPtr works properly in almost every other gtk s/r.  For instance, also in an example given above with gtk_list_child_position () where it works correctly and it too has "GTK_IS_LIST (list)" in the c source code.  clearly most of Gtk sees/receives that c_Ptr as a "list".  

Why Gtk does not "see things correctly (or even consistently)" for one or two s/r's is beyond us.

However, we did some experiments on the Fortran side to see what we would get by implementing the "guts" of the c-side functions.  For example, the `gtk_list_select_item()` s/r on the c-side employs `g_list_nth()` to get the "nth child", and returns a pointer.  If we implement 

    Temp_cPtr = g_list_nth( ListBox_cPtr, 1_c_Int);  ! which is in gtklist.c

it returns Temp_cPtr = 1 (the list had 23 items)!  Surely, that can't be correct.  

With:

    Temp_cPtr = `g_list_nth( ListBox_cPtr, 0_c_Int);

it returns Temp_cPtr = 40145056, which at least "looks like" a pointer value, unfortunately it is the pointer value for ListBox_cPtr itself, and not the "first (i.e. 0th)" element in the list  (the list had 23 items).

Any other choice for "n" that we tested SIGSEGV'd  (again, the list had 23 items).

It also explains why `gtk_list_select_item()` SIGSEGV's, since it relies on `g_list_nth()`, but it does not check for the validity of the g_list_nth return internally, and thus passes nonsense to an internal call to `gtk_list_select_child()` (i.e. `gtk_list_select_item()` is just a wrapper for `gtk_list_select_child()`).  

At this point it is starting to get "a little too Gtk deep" for us, but it does seem to suggest at least the "neighbourhood" in GLib/Gtk where things go wrong in terms of Gtk-Fortran interop.


We look forward to your thoughts.


ASIDE:  Although more properly the subject of a separate "issue".  It seems clear that one must be extremely careful with respect to which gtk win binaries, or which/how one builds their own win binaries (as we do), since "unknown differences" can cause one set of binaries to be (much) less stable compared to another ...  it is no wonder why, for example, GNumerics (written in Gtk) had announced that they will no longer create a Windows version, since they found Gtk to be much too unstable on Windows ... this does seem a little ominous ... does this sort of "stability" issue spell death for Gtk/Gtk-Fortran on Windows ? ... we'll do some more testing, but we are concerned.
